### ISSUE TITLE
Using plugin.google.maps.Map.getMap with 'target' : latLngBounds

### ISSUE THREAD

**Issue by johnrobertcobbold** [mapsplugin/cordova-plugin-googlemaps#647]
As far as I can see, it does not seem to be possible to initialise a map with getMap by using the target field with a latLngBounds, Only way seems to be to give a 'latLng' and a 'zoom'.

I am calculating a latLngBounds before displaying the map and would like to use that to automatically get the correct center and zoom level. I can of course get the latLngBounds' center and use that as the latLng but I cannot (as far as I can see) calculate the correct zoom level.

My current workaround is to calculate that latLngBounds' center and use an arbitrary high zoom level. When MAP_READY fires, I then use animateCamera with does accept a latLngBounds to zoom out to the correct zoom level. This of course implies a whole animation which I could do without. Also tried moveCamera but the map then displays briefly before suddenly shifting to the correct location which looks pretty bad.

In a nutshell, it would be great if we could use getMap with a latLngBounds

**Comment by johnrobertcobbold** [comment]
Hi @Hirbod 

Did some further research (http://stackoverflow.com/questions/6048975/google-maps-v3-how-to-calculate-the-zoom-level-for-a-given-bounds)  and it might not be possible to fit to bounds before the map is initialised.

As such, I adapted a function provided on the above link to calculate the correct zoom level, which I am using as such :

(1) Calculate the LatLngBounds
(2) Get its center with getCenter()
(3) Calculate the relevant zoom level using this function (adapted to use this plugin's LatLngBounds properties)  :

'    function getBoundsZoomLevel(bounds, mapDim) {
    var WORLD_DIM = { height: 256, width: 256 };
    var ZOOM_MAX = 21;

    function latRad(lat) {
        var sin = Math.sin(lat * Math.PI / 180);
        var radX2 = Math.log((1 + sin) / (1 - sin)) / 2;
        return Math.max(Math.min(radX2, Math.PI), -Math.PI) / 2;
    }

    function zoom(mapPx, worldPx, fraction) {
        return Math.floor(Math.log(mapPx / worldPx / fraction) / Math.LN2);
    }

    var ne = bounds.northeast;
    var sw = bounds.southwest;

    var latFraction = (latRad(ne.lat) - latRad(sw.lat)) / Math.PI;

    var lngDiff = ne.lng - sw.lng;
    var lngFraction = ((lngDiff < 0) ? (lngDiff + 360) : lngDiff) / 360;

    var latZoom = zoom(mapDim.height, WORLD_DIM.height, latFraction);
    var lngZoom = zoom(mapDim.width, WORLD_DIM.width, lngFraction);

    return Math.min(latZoom, lngZoom, ZOOM_MAX);
}'

(4) Init map as usual with calculated latLng and zoom

Works very nicely.

**Comment by Hirbod** [comment]
Hi @johnrobertcobbold 

thanks a lot! Could you provide a better example for other users please? A minimal JS-Code how you pass everything? Thanks!

**Comment by johnrobertcobbold** [comment]
Thanks @Hirbod but credits go to the guy @ http://stackoverflow.com/questions/6048975/google-maps-v3-how-to-calculate-the-zoom-level-for-a-given-bounds who posted the *getBoundsZoomLevel* function. I just adapted 2-3 small things to make it work with *phonegap-googlemaps-plugin*

As for WORLD_DIM : "At zoom zero, the whole world map is 256x256 pixels, and zooming each level doubles both width and height. So after a little algebra we can calculate the zoom as follows, provided we know the map's width in pixels. Note that because longitude wraps around, we have to make sure the angle is positive."

// start with an array of points that I want to zoom on when opening the map
`var arrayOfLatLng = [ new plugin.google.maps.LatLng(), new plugin.google.maps.LatLng(), ... ];`

// create bounds
`var bounds = new plugin.google.maps.LatLngBounds(arrayOfLatLng);`

// get their center
`var center = bounds.getCenter();`

// get the right zoom using the bounds and the dimensions of the map element in pixels
`var mapElement = document.getElementById('map');
var mapDimensions = { height: mapElement.offsetHeight ,width: mapElement.offsetWidth};
var zoom = getBoundsZoomLevel( bounds, mapDimensions };`

// initialize map as usual
`var map = plugin.google.maps.Map.getMap(
mapElement,
{ ..., 'camera': { 'latLng': center, 'zoom': zoom, ... } }
);`

LatLngBounds Class has a getCenter method. Perhaps you could consider adding a getZoom method? It could for example either take :
(1) one parameter (bounds) and automatically retrieve the map dimensions
(2) two parameters (bounds and dimensions). Useful if you are overlaying some html elements at the top / bottom of the maps and need a smaller zoom to avoid some points from being displayed below those html elements, Might be the most flexible one.

**Comment by Hirbod** [comment]
Thanks. 1.3.4 will have getZoom and getTilt for the map, should be enough, no?
Btw, I've just reformatted you post to increase readability.

**Comment by Hirbod** [comment]
Or did you mean I should implement your function into core?

```js
map.getBoundsZoomLevel(bounds.getCenter(), dimension[optional]);
```

**Comment by johnrobertcobbold** [comment]
This functions takes a *LatLngBounds* and calculates the maximum zoom level which the map (that has a specific width and height in pixels) can use to show that LatLngBounds. Combined with getCenter() which the LatLngBounds class already has, it allows to perfectly move and zoom the map camera right on a LatLngBounds.

**Comment by Hirbod** [comment]
Alright, I will consider to add this into the core. As the js-class already know the map element, we could drop the second parameter. I could retrieve offsetHeight and offsetWidth automatically.

**Comment by johnrobertcobbold** [comment]
There is one reason why you might want allow users to input a specific height so I drafted a quick sketch below. If you are displaying overlays at the top and/or bottom of the screen, the 'usable height' for the correct zoom calculation is not the map height anymore but a smaller height :

![googlemaps](https://cloud.githubusercontent.com/assets/307565/10487002/8006868e-72c4-11e5-8dd9-c4112e51257b.jpg)

In my case, the "usable height" is map height - (height + top margin of top overlay ) - ( height + bottom margin of bottom overlay)

**Comment by Hirbod** [comment]
So, I will just determine the dimensions automatically, when no dimensions-object was submitted.

**Comment by johnrobertcobbold** [comment]
Sounds good, hope it helps others

**Comment by nicooprat** [comment]
Was just looking for this feature, many thanks, works great. Would be useful to be able to pass a `LatLngBounds` object to the `zoom` parameter (instead of an integer) to indicate that we need to fit these bounds. For example something like :

````
map.animateCamera({
    target: new plugin.google.maps.LatLngBounds(points),
    zoom: new plugin.google.maps.LatLngBounds(points)
});
````

**Comment by johnrobertcobbold** [comment]
Glad it helped you. animateCamera does take a latLngBounds though btw. Was not able to use a latLngBounds when initializing the map with plugin.google.maps.Map.getMap, that was the issue on my side.

**Comment by Hirbod** [comment]
@johnrobertcobbold I'm very busy these days. Would you mind to provide a PR for this?

**Comment by wf9a5m75** [comment]
Google Maps SDK for iOS provides `fitBounds` feature, but Google Maps Android api v2 does not provide it.

**Comment by wf9a5m75** [comment]
Hi everyone, I think you don't see this page, but `map.animateCamera()` accepts `LatLngBounds` or `Array&lt;LatLng&gt;` also.
https://github.com/mapsplugin/cordova-plugin-googlemaps/wiki/CameraPosition#cameraposition-class-reference

```js
map.animateCamera({
  target: [
    new plugin.google.maps.LatLng(...),
    new plugin.google.maps.LatLng(...),
    new plugin.google.maps.LatLng(...)
  ],
  zoom : 10 // <-- ignored if you set target is Array<LatLng> or LatLngBounds
}, function() {
 ...
});
```

For the first question, `map.getMap()` does not accept `Array&lt;LatLng&gt;` or `LatLngBounds` certainly.
However you can use `map.moveCamera()` which changes the camera view without animation.
I will consider passing `Array&lt;LatLng&gt;` or `LatLngBounds` as getMap() options.

**Comment by wf9a5m75** [comment]
This request has been accepted, and implemented.
You can do like this: 

```js
map = plugin.google.maps.Map.getMap(div, {

  camera: {
    target : [
      new plugin.google.maps.LatLng(41.79883, 140.75675),
      new plugin.google.maps.LatLng(41.799240000000005, 140.75875000000002),
      new plugin.google.maps.LatLng(41.797650000000004, 140.75905),
      new plugin.google.maps.LatLng(41.79637, 140.76018000000002),
      new plugin.google.maps.LatLng(41.79567, 140.75845),
      new plugin.google.maps.LatLng(41.794470000000004, 140.75714000000002),
      new plugin.google.maps.LatLng(41.795010000000005, 140.75611),
      new plugin.google.maps.LatLng(41.79477000000001, 140.75484),
      new plugin.google.maps.LatLng(41.79576, 140.75475),
      new plugin.google.maps.LatLng(41.796150000000004, 140.75364000000002),
      new plugin.google.maps.LatLng(41.79744, 140.75454000000002),
      new plugin.google.maps.LatLng(41.79909000000001, 140.75465),
      new plugin.google.maps.LatLng(41.79883, 140.75673)
    ],
    'tilt': 60,
    'bearing': 50
  }
});
```

The mapOptions use `target` property instead of `latLng` property.
If you set `latLng` property, the `target` property is replaced automatically.

If you set `Array<LatLng>`, or `LatLngBounds` to the `target` property,
the `zoom` property is ignored, because the zoom level is set automatically to fit the specified bounds.

`map.animateCamera()` and `map.moveCamera()` are also available.

----

**REMEMBER**

The native Google Maps APIs do not support moving the camera position with `bounds`, `tilt` and `bearing`.
Hence, the plugin code executes multiple camera position moving.

For example, these two codes are the same meaning.

*old style*
```js
map.animateCamera({
  target: [
    new plugin.google.maps.LatLng(...),
    new plugin.google.maps.LatLng(...),
    new plugin.google.maps.LatLng(...)
  ]
}, function() {
  map.moveCamera({
    tilt: 30,
    bearing: 60
  }, function() {
    //done
  });
});
```

*new style*
```js
map.animateCamera({
  target: [
    new plugin.google.maps.LatLng(...),
    new plugin.google.maps.LatLng(...),
    new plugin.google.maps.LatLng(...)
  ],
  tilt: 30,
  bearing: 60
}, function() {
  //done
});
```
